### ISSUE TITLE
Request: methods for converting between different precisions

### ISSUE THREAD

**Issue by lukehutch** [JOML-CI/JOML#201]
There are a few methods for converting between `-f` and `-d` types, but the coverage is inconsistent. For example, `Quaternionf#setAxisAngle(float, float, float, float)` and `Quaternionf#setAxisAngle(double, double, double, double)` both exist, but even though there is a method `Quaternionf#set(AxisAngle4d)` and `Quaternionf#set(AxisAngle4f)`, and a `Quaternionf#set(float, float, float, float)`, there is no `Quaternionf#set(double, double, double, double)`.

In general there is no simple way to convert directly from `Quaternionf` to `Quaterniond` or vice versa, or `Vector3f` to `Vector3d` or vice versa.

It would also be nice to be able to call `Quaterniond#transform(Vector3f)`, since you generally want to keep quaternion math as accurate as possible (`-d`), but you may want to save space in storing vertices (`-f`). Having to convert from `Vector3f` to `Vector3d` to transform by a `Quaterniond` (or transforming the `Quaterniond` to a `Quaternionf`) makes the API harder to use. (Probably for `Quaterniond#transform(Vector3f)`, the return type should be a `Vector3f`, not a `Vector3d`, to save on yet another conversion the user would need to perform on return, to get the transformed coordinates back to the same precision type that they're already working with.)

**Comment by httpdigest** [comment]
I agree that between the single-precision and double-precision classes a more consistent API is needed when transforming between them in various operations. In fact, this is so systematic that this calls for an automated generator approach to generate float/double classes from a single template. Such a generator simply does not exist (yet) and therefore we need to handle those things on an as-needed basis and add the methods manually.
Can you list the exact methods/constructors you would like to have?
From your description I read the following:
- `Quaternionf(Quaterniond)`
- `Quaternionf.set(double, double, double, double)`
- `Quaterniond.get(Quaternionf)`
- `Quaterniond.transform(Vector3f)`

**Comment by lukehutch** [comment]
Those are the ones I most immediately needed, and had to hack around by casting individual quaternion components to make a new intermediate quaternion, yes. But let's make these symmetrical at least, so that the inverse is also covered:

* `Quaterniond(Quaternionf)`
* `Quaterniond.set(float, float, float, float)`
* `Quaternionf.get(Quaterniond)`
* `Quaternionf.transform(Vector3d)`

Also probably 

* `Quaterniond.set(Quaternionf)`
* `Quaternionf.set(Quaterniond)`

I totally understand that maintaining the cross product of all possible options would be tedious...

**Comment by httpdigest** [comment]
In the latest 1.9.20-SNAPSHOT build you'll find various added methods/constructors.

**Comment by lukehutch** [comment]
Thank you!

**Comment by lukehutch** [comment]
Some more suggested methods -- the ability to set a pure quaternion from a vector, i.e.

* `Quaterniond.set(Vector3d)` and `new Quaterniond(Vector3d)`
* `Quaternionf.set(Vector3f)` and `new Quaternionf(Vector3f)`
* `Quaterniond.set(Vector3f)` and `new Quaterniond(Vector3f)`
* `Quaternionf.set(Vector3d)` and `new Quaternionf(Vector3d)`

**Comment by lukehutch** [comment]
Some more suggestions:

* `Quaterniond.setFromAxisAngle(double angle, double x, double y, double z)`

...and the other three versions (to include float and mixed precision).

Incidentally the parameter order for the `AxisAngle` methods is unfortunate, because the order is actually angle first, then axis -- and also because the `Quaternion` methods take the vector first, then `w`, but `AxisAngle` taxes the angle first, and the vector last.

**Comment by httpdigest** [comment]
> * `Quaterniond.setFromAxisAngle(double angle, double x, double y, double z)`
> 
> ...and the other three versions (to include float and mixed precision).

This exists already: [Quaterniond setAngleAxis(double angle, double x, double y, double z)](https://joml-ci.github.io/JOML/apidocs/org/joml/Quaterniond.html#setAngleAxis(double,double,double,double))

**Comment by lukehutch** [comment]
Sorry, not sure how I missed that!

**Comment by httpdigest** [comment]
Probably because of the inconsistent naming - when comparing with set**From**Normalized(Matrix) . :)

**Comment by lukehutch** [comment]
In general I have found it hard to find methods, I have had to just guess at names until I figured out a method existed, or was pretty sure it didn't. Maybe JOML docs should have a table or graph saying "if you have X and want Y, call this method", for all combinations. Actually producing that in graph form would help find places where methods were missing, due to broken symmetry.

On my last point a couple of comments ago -- I know it's way too late to change the parameter order for quaternions, or to change `AxisAngle` to `AngleAxis`, but if you ever do a JOML 2.0, I strongly recommend making names and parameter orders consistent, and making angle/axis representation parallel quaternion parameter order. (`setFromAngleAxis` is another example of naming and parameter order inconsistency, vs. `AxisAngle`...) If you have a list of JOML 2.0 ideas, maybe add those things to your list!
